File: python/0004_median_of_two_sorted_arrays.py
def findMedianSortedArrays(nums1, nums2):
    def find_k(sub1, sub2, k):
        if len(sub1) == 0:
            return sub2[k - 1]
        if len(sub2) == 0:
            return sub1[k - 1]
        if k == 1:
            return min(sub1[0], sub2[0])
        pos1 = min(len(sub1), k // 2)
        pos2 = min(len(sub2), k // 2)
        if sub1[pos1 - 1] < sub2[pos2 - 1]:
            return find_k(sub1[pos1:], sub2, k - pos1)
        else:
            return find_k(sub1, sub2[pos2:], k - pos2)

    def find_k2(ind1, ind2, k):
        if ind1 == len(nums1):
            return nums2[k + ind2 - 1]
        if ind2 == len(nums2):
            return nums1[k + ind1 - 1]
        if k == 1:
            return min(nums1[ind1], nums2[ind2])
        pos1 = min(len(nums1), k // 2 + ind1)
        pos2 = min(len(nums2), k // 2 + ind2)
        if nums1[pos1 - 1] < nums2[pos2 - 1]:
            return find_k2(pos1, ind2, k - (pos1 - ind1))
        else:
            return find_k2(ind1, pos2, k - (pos2 - ind2))

    if (len(nums1) + len(nums2)) % 2 != 0:
        return find_k(nums1, nums2, (len(nums1) + len(nums2)) // 2 + 1)
    else:
        return (find_k(nums1, nums2, (len(nums1) + len(nums2)) // 2 + 1) +
                find_k(nums1, nums2, (len(nums1) + len(nums2)) // 2)) / 2

    # if (len(nums1) + len(nums2)) % 2 != 0:
    #     return find_k2(0, 0, (len(nums1) + len(nums2)) // 2 + 1)
    # else:
    #     return (find_k2(0, 0, (len(nums1) + len(nums2)) // 2 + 1) +
    #             find_k2(0, 0, (len(nums1) + len(nums2)) // 2)) / 2


print(findMedianSortedArrays([2], [1, 3, 4]))
print(findMedianSortedArrays([1, 3, 5, 7, 9], [2, 4, 6, 8]))
print(findMedianSortedArrays([1, 2], [3, 4]))

File: cpp/0004_median_of_two_sorted_arrays.cpp
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        if (nums1.size() == 0) {
        
            int len = nums2.size();
            int mid = len / 2;
            if ((len % 2) == 0) 
                return (nums2[mid] + nums2[mid - 1]) / 2.0;
            else
                return nums2[mid];
        }

        if (nums2.size() == 0) {
            int len = nums1.size();
            int mid = len / 2;
            if (len % 2 == 0) 
                return (nums1[mid] + nums1[mid - 1]) / 2.0;
            else
                return nums1[mid];
        }

        int len1 = nums1.size();
        int len2 = nums2.size();
        int mid = (len1 + len2) / 2;
        int idx = 0, i = 0, j = 0;
        int last = 0, temp =0;

        while(idx <= mid) {
            idx ++;
            last = temp;
            if (i == len1) {
                temp = nums2[j];
                j++;
                continue;
            }

            if (j == len2) {
                temp = nums1[i];
                i++;
                continue;
            }

            if(nums1[i] <= nums2[j]) {
                temp = nums1[i];
                i++;
            } else {
                temp = nums2[j];
                j++;
            }
        }
        
        if((len1 + len2) % 2 == 0) {
            return (last + temp) / 2.0;
        } else {
            return temp;
        }
        
    }
};
